### PR TITLE
feat: make sws reload only restart changed sw

### PR DESCRIPTION
**Pull request by xbsura**
Ref: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/448

make when vmagent reload config, only changed sw restart

### CODE REVIEW

**Comment by xbsura** [comment]
I found some runSD*** function are so similar, and it seems all support reload, so I add a function replace all of them, and support part restart at one time

File: lib/promscrape/scraper.go [hunk 1]
```diff
@@ -390,25 +401,25 @@ func equalLabel(a, b *prompbmarshal.Label) bool {
...
-		var sw scrapeWork
-		sw.Config = *cfg
-		sw.ReadData = c.ReadData
-		sw.PushData = pushData
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			sw.run(stopCh)
-		}()
+func hashScrapeWork(sw ScrapeWork) (string, error) {
```
Comment (xbsura) [inline lib/promscrape/scraper.go:404]: bug here, as bson encode can not get un export value, some change such as cert will not be watched

File: go.mod [hunk 2]
```diff
@@ -20,6 +20,7 @@ require (
 	github.com/valyala/gozstd v1.7.0
 	github.com/valyala/histogram v1.0.1
 	github.com/valyala/quicktemplate v1.4.1
+	go.mongodb.org/mongo-driver v1.3.2
```
Comment (valyala) [inline go.mod:23]: I'm afraid this violates the following principle - https://medium.com/@valyala/stripping-dependency-bloat-in-victoriametrics-docker-image-983fb5912b0d :(

What is the reason to include it here?

File: lib/httpserver/httpserver.go [hunk 3]
```diff
@@ -32,8 +32,9 @@ var (
 	metricsAuthKey   = flag.String("metricsAuthKey", "", "Auth key for /metrics. It overrides httpAuth settings")
 	pprofAuthKey     = flag.String("pprofAuthKey", "", "Auth key for /debug/pprof. It overrides httpAuth settings")
 
-	disableResponseCompression  = flag.Bool("http.disableResponseCompression", false, "Disable compression of HTTP responses for saving CPU resources. By default compression is enabled to save network bandwidth")
-	maxGracefulShutdownDuration = flag.Duration("http.maxGracefulShutdownDuration", 7*time.Second, "The maximum duration for graceful shutdown of HTTP server. "+
+	disableResponseCompression = flag.Bool("http.disableResponseCompression", false, "Disable compression of HTTP responses for saving CPU resources. By default compression is enabled to save network bandwidth")
+	// 7 is too small for online, and no argument can change this, make it bigger
+	maxGracefulShutdownDuration = flag.Duration("http.maxGracefulShutdownDuration", 120*time.Second, "The maximum duration for graceful shutdown of HTTP server. "+
```
Comment (valyala) [inline lib/httpserver/httpserver.go:37]: 120 seconds timeout is too big for graceful shutdown of http server, since this may increase VM service outage on restart from a few seconds to up to 2 minutes. I improved error message for graceful shutdown timeout in the commit 01c17092e1a9f260ca77c90936efab671535c10a , so users could adjust `-http.maxGracefulShutdownDuration` flag value if they see this error.

File: lib/netutil/conn.go [hunk 4]
```diff
@@ -27,19 +27,19 @@ type connMetrics struct {
 }
 
 func (cm *connMetrics) init(group, name, addr string) {
-	cm.readCalls = metrics.NewCounter(fmt.Sprintf(`%s_read_calls_total{name=%q, addr=%q}`, group, name, addr))
-	cm.readBytes = metrics.NewCounter(fmt.Sprintf(`%s_read_bytes_total{name=%q, addr=%q}`, group, name, addr))
-	cm.readErrors = metrics.NewCounter(fmt.Sprintf(`%s_errors_total{name=%q, addr=%q, type="read"}`, group, name, addr))
-	cm.readTimeouts = metrics.NewCounter(fmt.Sprintf(`%s_read_timeouts_total{name=%q, addr=%q}`, group, name, addr))
+	cm.readCalls = metrics.GetOrCreateCounter(fmt.Sprintf(`%s_read_calls_total{name=%q, addr=%q}`, group, name, addr))
```
Comment (valyala) [inline lib/netutil/conn.go:30]: Could you comment why this change is needed?

File: lib/httpserver/httpserver.go (hunk 3 above)
Comment (xbsura) [reply lib/httpserver/httpserver.go:37]: ok

File: go.mod (hunk 2 above)
Comment (xbsura) [reply go.mod:23]: go build -ldflags "-s -w" main.go, vmagent-prod size keeps 12M, same with last version prod

I am trying to do a hash for sw, to check whether it changes, of course I can make a []byte manually using sew struct, but seems a little ugly, so I use bson encode

bson comes from mongo-driver, I will think about make it more light

File: lib/netutil/conn.go (hunk 4 above)
Comment (xbsura) [reply lib/netutil/conn.go:30]: emmm, this is not so needed here, I will make another pr if necessary

In my option, GetOrCreateCounter is more safe than NewCounter, as GetOrCreateCounter can run multi times and NewCounter will panic if do more than one time

**Comment by xbsura** [comment]
> This pull request list looks too big. Could you try making it less intrusive or split it into multiple smaller pull requests?

does it mean the whole pr?

ok, I will split it to two minor pr, one is merge all runSw to one, beacuse almost all the code is same, and the other is make sw worker not restart if not necessary

because this feature need to control stop for all sw, but not using a single channel stop all sw, so maybe the change code is a little many

I will try change less code

thanks for your reply, this feature is important for us, as our env changes config file multi times per minute, if restart all sws, metrics lost is very heavy

**Comment by valyala** [comment]
> this feature is important for us, as our env changes config file multi times per minute, if restart all sws, metrics lost is very heavy

I see and fully agree with the motivation behind this pull request. Thanks for trying to solve the issue!

**Comment by valyala** [comment]
@xbsura , I pushed the commit 3b234d82e541395628951db1b70f0d5079ae9918 , which is based on the ideas from this PR. Thanks for the ideas and initial implementation attempts!

File: lib/httpserver/httpserver.go (hunk 3 above)
Comment (valyala) [reply lib/httpserver/httpserver.go:37]: FYI, there was a bug in `lib/httpserver`, which could increase graceful shutdown duration for http server to more than a minute. This should be fixed in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/787fcfba0ce361b728fb639496dd24994f4e7cd9 . Now typical graceful shutdown duration should take less than 5 seconds. The duration can be longer if slow requests are processed during the shutdown.

File: go.mod (hunk 2 above)
Comment (valyala) [reply go.mod:23]: I see. It is still better to write or copy-paste small code snippets from external package instead of importing it. See https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s and https://news.ycombinator.com/item?id=14903059 for details.

File: lib/netutil/conn.go (hunk 4 above)
Comment (valyala) [reply lib/netutil/conn.go:30]: `metrics.NewCounter` is used intentionally in places where it is expected that they are executed only once. This allows immediately catching the bug via `panic` when the code is executed multiple times.
